rayon-core/src/scope/mod.rs

```diff
@@ -536,12 +530,14 @@ impl<'scope> ScopeBase<'scope> {
     /// appropriate.
     ///
     /// Unsafe because it must be executed on a worker thread.
-    unsafe fn complete<FUNC, R>(&self, owner_thread: &WorkerThread, func: FUNC) -> R
+    unsafe fn complete<FUNC, R>(&self, func: FUNC) -> R
     where
         FUNC: FnOnce() -> R,
     {
         let result = self.execute_job_closure(func);
-        self.steal_till_jobs_complete(owner_thread);
+        in_worker(|owner_thread, _| {
```

Comment:
It seems unfortunate to inject yet another job just to wait for completion. Maybe we could add another latch type that acts like a combination of `CountLatch` and `LockLatch`, raising the `Condvar` when the count is zero. Then `job_completed_latch` would need to be an enum of either `CountLatch` or this new thing, depending on what context it's called from.

Reply:
This doesn't matter for rustc, since it always runs in a worker.